python/ray/data/_internal/compute.py

```diff
@@ -359,8 +394,50 @@ def map_block_nosplit(
                     block_indices[ref] = len(blocks_in)
                     tasks_in_flight[worker] += 1
 
+                # Try to scale down if there's no more work to do and pending/idle
+                # workers.
+                if not blocks_in:
+                    should_update = False
+                    if tasks_in_flight[worker] == 0:
+                        should_update = True
+                        # No more work to do and worker is idle, so we terminate the
+                        # worker.
+                        ready_workers.remove(worker)
+                        workers.remove(worker)
+                        # Explicitly terminate the actor to expedite cleanup.
+                        ray.kill(worker)
```

Comment:
Might be better to go with graceful termination here so user-defined finalizers run and so the termination is a bit more "clean"? The worker is guaranteed to not have any outstanding tasks since `tasks_in_flight[worker] == 0`, so this shouldn't matter much in practice AFAIK.